lldb/tools/lldb-server/lldb-platform.cpp

```diff
@@ -150,12 +153,17 @@ static void client_handle(GDBRemoteCommunicationServerPlatform &platform,
   printf("Disconnected.\n");
 }
 
-static GDBRemoteCommunicationServerPlatform::PortMap gdbserver_portmap;
-static std::mutex gdbserver_portmap_mutex;
-
 static void spawn_process_reaped(lldb::pid_t pid, int signal, int status) {
-  std::lock_guard<std::mutex> guard(gdbserver_portmap_mutex);
-  gdbserver_portmap.FreePortForProcess(pid);
+  if (g_single_pid != LLDB_INVALID_PROCESS_ID && g_single_pid == pid) {
+    // If not running as a server and the platform connection is closed
+    if (!g_terminate) {
+      g_terminate = true;
```

Comment:
We need this callback only in case of non-server mode and tcp protocol.
Note this callback is called from the MonitorChildProcessThread. 
See the issue #101475. We must be very careful with this callback and we cannot change any arguments. 
I have removed g_terminate.

Reply:
Yes, we need to be very careful, but I don't think this is how being careful looks like. `SetMonitorProcessCallback` takes a std::function, so we can add (or remove) as many arguments as we like.

Reply:
Ok, I see what you meant with https://github.com/llvm/llvm-project/issues/101475 now. That said I don't think this solves anything, cause the same thing could now happen with the mainloop pointer (if it is destroyed just as this function is exiting). The way I'd ideally solve this is by ensuring that the main loop does not terminate (which means the main loop and the platform objects will not be destroyed) until this function is called.

Reply:
Note this callback is called from an uncontrollable thread and any pending arguments in the std::function wrapper may be invalid during the process destroying procedure.
I have updated the logic around g_server and removed <atomic> and g_single_pid at all.

Reply:
I'm not particularly happy with the design of this callback, but that doesn't mean it can't be used safely. The "uncontrollable" aspect means it is up to you (its user) to ensure the callback does not access any arguments after they are destroyed. It doesn't matter if they're local or global or how they're passed into the callback. The new implementation is better (because it's easier to reason about), but it still accesses the global main loop object, which will get destroyed some time after `main` returns. So, the question here is "is there anything ensuring that does not happen?" and not "is mainloop a global?". If there is (for example, because the main thread waits for the callback to call `RequestTermination()`), then I'm pretty this could be refactored to avoid the global. If there isn't, then this is a problem even though the variable is global.

Reply:
There are a lot of discussions around `Detached threads accessing global or static objects` in the internet.
Detached threads are bad and the design of this callback is bad.
But if GDBRemoteCommunicationServerPlatform::g_spawned_pids is static (now) it reduces the chances of a crash to 0.(0)1%.
We have no handle of this thread. So the static global g_main_loop is the safest way and g_main_loop must be destroyed (almost) last. No 100% guarantee. But it is the better solution w/o redesigning the child process monitoring.
Ultimately we can just remove `fixes #101475` in the description because this patch improves the situation but do not fix it 100%.

Reply:
I think I know the problems with detached threads pretty well (remember how I said I found bugs in lldb's dependencies -- [one of them](https://sourceware.org/bugzilla/show_bug.cgi?id=19951) is in glibc's pthread_detach). And I'm not interested in changing the design of the monitoring callback. It's detached and it's going to stay detached for a long time to come.

The thing I'm questioning is whether the code you are running inside the callback has to be detached (because the fact that the outer callback machinery is detached doesn't mean that the callback itself has to be detached).

The problems with detached threads are a runtime property. If a thread *accesses* something after it has been destroyed, you have a bug. I emphasised "access" because the access has to happen. Mere existence of a pointer/reference that would permit that access is not in itself a problem. So code like
```
int main() {
  std::condition_variable cv;
  std::thread([&]{cv.notify_one();}).detach();
  cv.wait();
  return;
}
```
is safe because even though the detached thread holds a reference to the local object (and may continue to hold it after its destroyed), the program ensures that the (only) usage of that reference happens while the object is alive. If I removed `cv.wait()` then we'd have a problem.

Why I'm mentioning this? Because this is exactly the pattern we have (or one that I think we should have) here, just instead of `cv.wait()` we have `mainloop.Run()` and `cv.notify()` is replaced by `mainloop.RequestTermination()`

If we can make sure things work this way, we're safe, or at least the (0.1%) problems are "not our problem". If we can't undetach this code, then we're exacerbating the problem, because even if someone wanted to make the undetach the monitoring callback machinery, he could make that work because the code inside it would be detached.

Reply:
The child monitor callback is used in 2 places:

* lldb-platform.cpp: In case of non-server mode we need to exit after disconnecting (when the child process handling the platform connection exited). In theory we should be in the middle of MainLoop except the case when an assert happened somewhere. So access g_main_loop is safe here. But the safest way is to use a terminate flag and a dummy connection instead of accesing g_main_loop. We can pass a reference to MainLoop as a callback argument but it changes nothing. A global variable is much clear here.

* The child process handling the platform connection uses the single instance of GDBRemoteCommunicationServerPlatform, which uses a list of PIDs g_spawned_pids. This process will exit after disconnecting and the child process monitor callback is most dangerous point. Any processes may exit at any time. We need the PIDs list only for Handle_qKillSpawnedProcess response. It would be better to try to kill the process anyway and just check somehow that this PID is still alive. I think the best solution here is the static g_spawned_pids. Any other solutions are out the scope of this patch.